website/content/specification/syntax/constraints.md

```diff
@@ -14,27 +14,79 @@ TODO: P3: Address issue https://github.com/usnistgov/metaschema/issues/325
 
-If an `allowed-values` constraint does not have the `allow-other` attribute defined, the default is `allow-other="no"`, resulting in strict validation where the only valid values are those in the list.
+The *source* of each allowed `<allowed-values>` constraint in the *target set* is identified as one of the following sources:
 
-Within `allowed-values` of a `constraint`, an `enum` element's `@value` attribute assigns the permissible value, while its data content provides documentation. For example:
+- **model:** The constraint is defined *in* a Metaschema module.
+- **external:** The constraint is defined *outside* a Metaschema module, e.g. external constraints.
+
+The *target set* of `<allowed-values>` constraints is verified for correctness using the `@extension` attribute on each set member.
+
+Once the *target set* of `<allowed-values>` constraints is validated, the `@allow-other` attribute is used to determine the *expected value set* for a given content value.
+
+The following subsections detail the processing requirements for the `@extension` and `@allow-other` attributes.
+
+### `@extension`
+
+For each `<allowed-values>` constraint, the `@extension` attribute MUST be one of the following values.
+
+- **`none`:** There can be no other matching `<allowed-values>` constraint for the same target value. This is the least permissive option.
+
+- **`model`:** (default) Multiple matching `<allowed-values>` constraints are allowed for the same target value as long as the constraints are defined in the same model. Constraints sourced from outside the model are not allowed.
+
+    All allowed-values constraints declared within a Metaschema model matching the same @target can be combined. If a matching constraint within the model has allow-other="no", then constraints declared externally from the model are not allowed. **This is the implicit default value if no `@extension` is provided.**
+
+- **`external`:** Multiple matching `<allowed-values>` constraints are allowed for the same target value, which can be sourced from within the model or externally through a set of external constraints.
+
+    All allowed-values constraints, declared within the model and externally through an extension, that match the same @target can be combined. This is the most permissive option.
+
+One of the following requirements MUST apply when processing a value's *target set* to validate it.
+
+1. The *target set* MUST contain a single `<allowed-values>` constraint with the `@extension` attribute value `none`.
+1. All `<allowed-values>` constraints in the *target set* MUST have the `@extension` attribute value `model` and originate from a *model* source.
+1. All `<allowed-values>` constraints in the *target set* MUST have the `@extension` attribute value `external` and originate from either a *model* or *extension* source.
+1. An error MUST be raised indicating the *target set* is invalid.
```

Comment:
> Hm. Does processing stop? Does a constraints implementation have the option of reporting such a discrepancy and continuing? If so, is it free to 'do its best'?

@wendellpiez  I talked with @aj-stein-nist about this. We need to be more clear around if an error is fatal or recoverable. We also need to be more clear around what is expected around user interaction, e.g. the user is notified.

> How many of these errors can be detected and prevented (or reported) at metaschema compile ('composition') time, rather than only at document parse-time? (Apologies if I have no ready answer to that yet.)

IMHO, preventing these errors in a different part of the processing is out of scope here.

This is about constraint processing, not parsing. We can't assume anything about implementation approaches. Not all implementations will have a "compile" step. The Java implementation doesn't have one for example.

Our responsibility in this part of the specification is to be clear about what constitutes an error and how an error should be handled.